### DISCUSSION TITLE
How to make registerUriHandler work

### DISCUSSION THREAD

**Discussion by futurist** [coder/code-server#3891] [Q&A]
How to handle protocol `vscode://` uri with existing extension?

From documentation of  `vscode.window.registerUriHandler()`:

> Registers a uri handler capable of handling system-wide uris. In case there are multiple windows open, the topmost window > will handle the uri. A uri handler is scoped to the extension it is contributed from; it will only be able to handle uris which are > directed to the extension itself.

I want to migrate an existing vscode plugin to code-server, but don't know how to get `handleUri` event work like below code:

```js
vscode.window.registerUriHandler({
    handleUri(uri:vscode.Uri) {
        // extension logic for `vscode://my-plugin/path`
    }
});
```

Local vscode can trigger `handleUri` message when open `vscode://my-plugin/path` from browser, how does code-server do similar like this? That is code-server already in browser, and how to notify my plugin with message passed from `vscode://my-plugin/path`, what's the right entry for this special URI?

**Comment by jsjoeio** [comment]
Oh this is a great question. I think @code-asher might have some tips/ideas

**Comment by code-asher** [comment]
I bet this hasn't been modified to work in the browser yet. 

We'll likely need to patch VS Code to use https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler

But this is definitely better done upstream.

**Comment by futurist** [comment]
Does `navigator.registerProtocolHandler` conflict with user's local VSCode? For example with the link `vscode://my-plugin/path`, should it be handled by browser or by VSCode? Also, user can have multiple instances of code-server.

Maybe code-server can have a special route it self to handle this message? Like below:

```
https://code-server-url/protocol-handler?uri=vscode://my-plugin/path
```

This added the benefit: there're no notification/confirmation when user open it, more smooth than VSCode.

**Reply by lg-nineoneninethreenine** [reply]
I think `vscode://my-plugin/path` is handled by local VSCode. And I think code-server has a special route.

**Reply by code-asher** [reply]
I do not think browsers allow handling `vscode://` anyway, at best we could do `web+vscode://` or `web+code-server://`.

See permitted schemes: https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler#permitted_schemes

The `/protocol-handler` idea is interesting! I think the only problem is how to get `code-server-url`. For example if GitHub had an "Open in code-server" button they would not have anything to put for the domain. So we will probably need something like `web+code-server://my-plugin/path`.

**Reply by ggydush-fn** [reply]
@code-asher was wondering if something like this is possible now? I love being able to clone directly from a GitHub button and would love for this functionality to be possible with code-server!

For example, the URL below clones a GitHub repo in your local Vscode.
```
vscode://vscode.git/clone?url=url
```

**Reply by code-asher** [reply]
No, I do not think anything has changed here yet.

**Reply by OCram85** [reply]
Could the `protocl_handlers` property in the PWA manifest help? (https://developer.mozilla.org/en-US/docs/Web/Manifest/protocol_handlers)

I'm also trying to make this work with a setup like this:
- test different urls within gitea 1.22.0 (https://blog.gitea.com/release-of-1.22.0/#rocket-customizable-open-with-ide-menu-29320)
- try to open code-server installed as PWA via Chrome on Windows
- testing workarounds like https://github.com/shengyou/vscode-handler

**Reply by jswhisperer** [reply]
> Could the `protocl_handlers` property in the PWA manifest help? (https://developer.mozilla.org/en-US/docs/Web/Manifest/protocol_handlers)
> 
> I'm also trying to make this work with a setup like this:
> 
> * test different urls within gitea 1.22.0 (https://blog.gitea.com/release-of-1.22.0/#rocket-customizable-open-with-ide-menu-29320)
> * try to open code-server installed as PWA via Chrome on Windows
> * testing workarounds like https://github.com/shengyou/vscode-handler

I know it's an old thread but just adjusting the manifest Works for me. You can test by overriding sources.
It must be served over https:// as per PWA spec.

<img width="961" alt="Screenshot 2024-11-06 at 10 20 55 PM" src="https://github.com/user-attachments/assets/b3edbbbf-6c94-4a9a-9cdd-f06f488c9a4f">
<img width="700" alt="Screenshot 2024-11-06 at 10 21 06 PM" src="https://github.com/user-attachments/assets/f7f2060b-9a50-44be-858e-b4b8835f7674">
<img width="480" alt="Screenshot 2024-11-06 at 10 21 12 PM" src="https://github.com/user-attachments/assets/5a92f5bd-f1f3-43d8-b2d4-cc3091d256df">
